### PR TITLE
net: coap: CoAP observe age now correctly wraps around

### PR DESCRIPTION
**Pull request by sibertdeclercq**
Fixes two bugs:
- When a notify is executed before any observers are added, the age is incremented from 0 to 1. When an observer is registered, the age is not 0 as expected,  causing the age to be 1 instead of the expected 2
- The check if a message is newer than the last received one is described in [RFC7641, section 3.4](https://datatracker.ietf.org/doc/html/rfc7641#section-3.4). Simply incrementing age will not comply to what the RFC describes once the value wraps around. This is now fixed

### CODE REVIEW

File: tests/net/lib/coap/src/main.c [hunk 1]
```diff
@@ -888,10 +908,10 @@ ZTEST(coap, test_observer_server)
 		0x51, 's', 0x01, '1', /* path */
 	};
 	uint8_t not_found_request_pdu[] = {
-		0x45, 0x01, 0x12, 0x34,
-		't', 'o', 'k', 'e', 'n',
+		0x45, 0x01, 0x12, 0x34, 
+		't', 'o', 'k', 'e', 'n', 
 		0x60, /* enable observe option */
-		0x51, 's', 0x01, '2', /* path */
+		0x51, 's',  0x01, '3', /* path */
```
Comment (rlubos) [inline tests/net/lib/coap/src/main.c:914]: Double space here, other formating issues above.

File: tests/net/lib/coap/src/main.c [hunk 2]
```diff
@@ -1110,14 +1130,14 @@ ZTEST(coap, test_build_options_out_of_order_0)
...
-		zassert_equal(expected_data_len, cpkt.offset, "Wrong offset");                     \
-		zassert_mem_equal(expected_data, cpkt.data, expected_data_len, "Wrong data");      \
+#define ASSERT_OPTIONS(cpkt, expected_opt_len, expected_data, expected_data_len)		\
+	do {											\
+		static const uint8_t expected_hdr_len = 9;					\
+		zassert_equal(expected_hdr_len, cpkt.hdr_len, "Wrong header length");		\
+		zassert_equal(expected_opt_len, cpkt.opt_len, "Wrong option length");		\
+		zassert_equal(expected_hdr_len + expected_opt_len, cpkt.offset, "Wrong offset");\
+		zassert_equal(expected_data_len, cpkt.offset, "Wrong offset");			\
+		zassert_mem_equal(expected_data, cpkt.data, expected_data_len, "Wrong data");	\
```
Comment (jukkar) [inline tests/net/lib/coap/src/main.c:1140]: This looks like unrelated cosmetic change if I managed to read the lines correctly. Please remove this from the commit if so. If you really want to change this, place the change to new commit in this PR.

Comment (rlubos) [inline tests/net/lib/coap/src/main.c:1140]: Why this change? Please don't introduce unnecessary modifications.

File: tests/net/lib/coap/src/main.c [hunk 3]
```diff
@@ -1771,4 +1791,45 @@ ZTEST(coap, test_transmission_parameters)
...
+	memcpy(data, valid_request_pdu, sizeof(valid_request_pdu));
+
+	r = coap_packet_parse(&req, data, sizeof(valid_request_pdu), options, opt_num);
+	zassert_equal(r, 0, "Could not initialize packet");
+
+	r = coap_handle_request(&req, server_resources, options, opt_num,
+				(struct sockaddr *)&dummy_addr, sizeof(dummy_addr));
+	zassert_equal(r, 0, "Could not handle packet");
+
+	// Forward time a bit, as not to run this 8 million time
```
Comment (rlubos) [inline tests/net/lib/coap/src/main.c:1817]: C-style `/* */` comments only please.

File: tests/net/lib/coap/src/main.c [hunk 4]
```diff
@@ -1771,4 +1791,45 @@ ZTEST(coap, test_transmission_parameters)
...
+	resource->age = INT32_MAX - 10;
```
Comment (rlubos) [inline tests/net/lib/coap/src/main.c:1818]: Is INT32_MAX actually correct assumption here, we've just introduced code that prevented 24-bit age overflow. Perhaps we should introduce a define for maximum age value and use it in code and test here?

**Review by jukkar — COMMENTED** [review]
Looks good, please fix the CI reported issues. Could you split the tests to separate commit, so 1st commit is the change and 2nd commit is the tests.

**Comment by rlubos** [comment]
Please use `net: coap: ` prefix for the commit title. The description you provid in the PR should be part of the commit message.

File: subsys/net/lib/coap/coap.c [hunk 5]
```diff
@@ -1750,7 +1750,8 @@ size_t coap_pendings_count(struct coap_pending *pendings, size_t len)
 }
```
Comment (jukkar) [inline subsys/net/lib/coap/coap.c:1750]: In the test commit, please do not mix code from `tests/net/...` and `subsys/net/....` in the same commit, but place the changes to `subsys/net/lib/coap/coap.c` to the first commit.
Also please change the test commit subject to `tests: net: coap: ....`

**Comment by jukkar** [comment]
There is some compliance issues reported
```
 Commit b8f49eb355:
4: UC4 Commit message body line exceeds max length (95>75): "* When a notify is executed before any observers are added, the age is incremented from 0 to 1."
5: UC4 Commit message body line exceeds max length (111>75): "When an observer is registered, the age is not 0 as expected, causing the age to be 1 instead of the expected 2"
6: UC4 Commit message body line exceeds max length (98>75): "* The check if a message is newer than the last received one is described in RFC7641, section 3.4."
7: UC4 Commit message body line exceeds max length (112>75): "Simply incrementing age will not comply to what the RFC describes once the value wraps around. This is now fixed"
```

File: subsys/net/lib/coap/coap.c [hunk 6]
```diff
@@ -1802,7 +1802,7 @@ struct coap_reply *coap_response_received(
 
 		age = coap_get_option_int(response, COAP_OPTION_OBSERVE);
 		/* handle observed requests only if received in order */
-		if (age == -ENOENT || is_newer(r->age, age)) {
+		if (age == -ENOENT || coap_age_is_newer(r->age, age)) {
```
Comment (jukkar) [inline subsys/net/lib/coap/coap.c:1805]: Please place this change to 1st commit.

**Review by pdgendt — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
